### ISSUE TITLE
Field `username` is set to null when getting users as admin

### ISSUE THREAD

**Issue by Rosiana18** [cloudfoundry/cloud_controller_ng#1574]
## Issue

When `cf curl`ing `/v3/users` as admin uaa user, `username` field is set to `null`, `origin` field is set to `null`, and `presentation_name` is set to guid.

## Context

We seem to find inconsistency with the user resource data using admin credentials. Whenever we hit `/v3/users`, we are seeing the `username` field set to `null`, `origin` field set to `null`, and `presentation_name` set to client ID but whenever we hit `/v3/users/:guid`, the `username`, `origin`, and `presentation_name` fields are the expected value. 

When `cf curl`ing `/v3/users` as a non-admin uaa user, data is expected with `username`, `origin` and `presentation_name` fields being valid.

We are only seeing this issue using the admin uaa user, behavior is expected as non admins.
The configuration for the users are the same default one as https://github.com/cloudfoundry/cf-deployment/blob/master/cf-deployment.yml#L593-L601 without the network.

## Steps to Reproduce

cf login -u admin
cf curl /v3/users

## Expected result

Getting user individually with the admin uaa client scope 
`cf curl "/v3/users/d1e8a1ca-994d-4992-b3f9-164c09eb9286"` returns the expected fields:

```
{
   "guid": "d1e8a1ca-994d-4992-b3f9-164c09eb9286",
   "created_at": "2015-01-21T23:54:45Z",
   "updated_at": "2019-06-06T01:25:32Z",
   "username": "admin",
   "presentation_name": "admin",
   "origin": "uaa",
   "metadata": {
      "labels": {},
      "annotations": {}
   },
   "links": {
      "self": {
         "href": "https://api.[domain]/v3/users/d1e8a1ca-994d-4992-b3f9-164c09eb9286"
      }
   }
}
```

Hence the expected result for `/v3/users` should be:
```
{
   "pagination": {
      "total_results": 1771,
      "total_pages": 1771,
      "first": {
         "href": "https://api.[domain]/v3/users?page=1&per_page=1"
      },
      "last": {
         "href": "https://api.[domain]/v3/users?page=1771&per_page=1"
      },
      "next": {
         "href": "https://api.[domain]/v3/users?page=2&per_page=1"
      },
      "previous": null
   },
   "resources": [
      {
         "guid": "d1e8a1ca-994d-4992-b3f9-164c09eb9286",
         "created_at": "2015-01-21T23:54:45Z",
         "updated_at": "2019-06-06T01:25:32Z",
         "username": "admin",
         "presentation_name": "admin",
         "origin": "uaa",
         "metadata": {
            "labels": {},
            "annotations": {}
         },
         "links": {
            "self": {
               "href": "https://api.[domain]/v3/users/d1e8a1ca-994d-4992-b3f9-164c09eb9286"
            }
         }
      }
   ]
}
```

## Current result

Getting users as admin uaa client returns `null` for `username` field, client ID for `presentation_name`, and `null` for `origin`.
```
{
   "pagination": {
      "total_results": 1771,
      "total_pages": 1771,
      "first": {
         "href": "https://api.[domain]/v3/users?page=1&per_page=1"
      },
      "last": {
         "href": "https://api.[domain]/v3/users?page=1771&per_page=1"
      },
      "next": {
         "href": "https://api.[domain]/v3/users?page=2&per_page=1"
      },
      "previous": null
   },
   "resources": [
      {
         "guid": "d1e8a1ca-994d-4992-b3f9-164c09eb9286",
         "created_at": "2015-01-21T23:54:45Z",
         "updated_at": "2019-06-06T01:25:32Z",
         "username": null,
         "presentation_name": "d1e8a1ca-994d-4992-b3f9-164c09eb9286",
         "origin": null,
         "metadata": {
            "labels": {},
            "annotations": {}
         },
         "links": {
            "self": {
               "href": "https://api.[domain]/v3/users/d1e8a1ca-994d-4992-b3f9-164c09eb9286"
            }
         }
      }
   ]
}
```

**Comment by aashah** [comment]
Hi,

I'm having trouble following what user is hitting `/v3/users/:guid`. Could you please provide more clear reproduction steps.

For reference, I tried the following two conditions

Using my CF Admin user, I expect `/v3/users/:guid` to be fully populated:

```
$ cf login -u admin
$ cf curl /v3/users/5482a214-c35a-4ec0-81ca-290b8825cee2
{
   "guid": "5482a214-c35a-4ec0-81ca-290b8825cee2",
   "created_at": "2020-04-24T22:03:55Z",
   "updated_at": "2020-04-24T22:03:55Z",
   "username": "admin",
   "presentation_name": "admin",
   "origin": "uaa",
   "metadata": {
      "labels": {},
      "annotations": {}
   },
   "links": {
      "self": {
         "href": "https://api.stitch-foe.capi.land/v3/users/5482a214-c35a-4ec0-81ca-290b8825cee2"
      }
   }
}
```

I then tried to use a UAA client with `uaa.admin`:

```
$ uaac client add foobar --name="test-cc-admin" --secret="secret" --authorities="uaa.admin" --authorized_grant_types="client_credentials"
$ cf auth --client-credentials foobar secret
$ cf curl /v3/users/5482a214-c35a-4ec0-81ca-290b8825cee2
FAILED
Error creating request:
Invalid auth token: Invalid Auth Token

$ uaac client update foobar --authorities "uaa.admin,cloud_controller.admin"
$ cf auth --client-credentials foobar secret
$ cf curl /v3/users/5482a214-c35a-4ec0-81ca-290b8825cee2
{
   "guid": "5482a214-c35a-4ec0-81ca-290b8825cee2",
   "created_at": "2020-04-24T22:03:55Z",
   "updated_at": "2020-04-24T22:03:55Z",
   "username": "admin",
   "presentation_name": "admin",
   "origin": "uaa",
   "metadata": {
      "labels": {},
      "annotations": {}
   },
   "links": {
      "self": {
         "href": "https://api.stitch-foe.capi.land/v3/users/5482a214-c35a-4ec0-81ca-290b8825cee2"
      }
   }
}
```

**Comment by amhuber** [comment]
The issue is that we're getting different data from querying all users vs. just a single user.  Here is an example:

`cf login -u admin`
`cf curl /v3/users`

Picking a random user from the data, it looks like this:

      {
         "guid": "e4a72b78-3b76-4f92-bf53-a154fe680a9b",
         "created_at": "2015-06-12T00:53:11Z",
         "updated_at": null,
         "username": null,
         "presentation_name": "e4a72b78-3b76-4f92-bf53-a154fe680a9b",
         "origin": null,
         "metadata": {
            "labels": {},
            "annotations": {}
         },
         "links": {
            "self": {
               "href": "https://api.bogus.com/v3/users/e4a72b78-3b76-4f92-bf53-a154fe680a9b"
            }
         }
      }

You can see that the username is "null" and the presentation name is a GUID.  But if I query for the user directly, it returns all the properties as expected:

`cf curl /v3/users/e4a72b78-3b76-4f92-bf53-a154fe680a9b`

      {
         "guid": "e4a72b78-3b76-4f92-bf53-a154fe680a9b",
         "created_at": "2015-06-12T00:53:11Z",
         "updated_at": null,
         "username": "valid.email@bogus.com",
         "presentation_name": "valid.email@bogus.com",
         "origin": "ldap",
         "metadata": {
            "labels": {},
            "annotations": {}
         },
         "links": {
            "self": {
               "href": "https://api.bogus.com/v3/users/e4a72b78-3b76-4f92-bf53-a154fe680a9b"
            }
         }
      }

**Comment by aashah** [comment]
Thanks,

Was definitely able to reproduce this. My new steps were:

```
$ for i in {1..1000}; do echo $i >> ids; done
$ cat ids.txt | xargs -P 20 -I{} -n1 /bin/bash -c "cf create-user user-{} password"
$ cf curl /v3/users
{
...
      {
         "guid": "76e85205-9069-4db9-800f-0dd3ab1db9cf",
         "created_at": "2020-04-28T20:41:18Z",
         "updated_at": "2020-04-28T20:41:18Z",
         "username": null,
         "presentation_name": "76e85205-9069-4db9-800f-0dd3ab1db9cf",
         "origin": null,
         "metadata": {
            "labels": {},
            "annotations": {}
         },
         "links": {
            "self": {
               "href": "https://api.jungle-prince.capi.land/v3/users/76e85205-9069-4db9-800f-0dd3ab1db9cf"
            }
         }
      }
   ]
}
$ cf curl /v3/users/76e85205-9069-4db9-800f-0dd3ab1db9cf
{
   "guid": "76e85205-9069-4db9-800f-0dd3ab1db9cf",
   "created_at": "2020-04-28T20:41:18Z",
   "updated_at": "2020-04-28T20:41:18Z",
   "username": "user-48",
   "presentation_name": "user-48",
   "origin": "uaa",
   "metadata": {
      "labels": {},
      "annotations": {}
   },
   "links": {
      "self": {
         "href": "https://api.jungle-prince.capi.land/v3/users/76e85205-9069-4db9-800f-0dd3ab1db9cf"
      }
   }
}
```

Looking at the `api/0` logs, I see the following:

```
{"timestamp":1588106922.798467,"message":"Failed to retrieve users from UAA: #<CF::UAA::BadResponse: received invalid response content or type>","log_level":"error","source":"cc.uaa_client","data":{"request_guid":"bca4dbae-f98f-4f28-7f60-24faeb6e7416::e726196c-1e1f-42d2-ad36-9411c1389c93"},"thread_id":46929056645560,"fiber_id":46929043747840,"process_id":8,"file":"/var/vcap/data/packages/cloud_controller_ng/125aa4448927f84dbe479f91ad85d3df30a0f5e6/cloud_controller_ng/lib/cloud_controller/uaa/uaa_client.rb","lineno":43,"method":"rescue in users_for_ids"}
```

I'm fairly certain the implementation of /v3/users is flawed. It takes a large list of guids, translates that into a SCIM filter (`id eq ABC or id eq DEF ...`) and sends that as a query param to UAA.

**Comment by reid47** [comment]
Thank you for reporting this, and thank you @aashah for your initial investigation! We are currently taking a deeper look and will hopefully get to the bottom of this. We are also able to reproduce it and we agree that there is probably something fishy about how the CC -> UAA communication is implemented.

**Comment by ChrisMtz** [comment]
Hello, this issue seems to be resolved in commit: https://github.com/cloudfoundry/cloud_controller_ng/commit/f3ac783214cbb9be26894e5cfc1faac5ddc07556

I have tested using cf curl and `V3/users` output matches `V3/user/:guid `output.
Can someone confirm and close issue?

**Comment by Gerg** [comment]
We believe this issue has been resolved by the commit linked above. I'm going to close this issue now, but feel free to re-open it if you are still having trouble.

Thanks,
Greg

**Comment by robachmann** [comment]
@Gerg in my case (CAPI 1.97.0, 3.87.0) `username` and `origin` are still set to `null` if I query with an admin user and request more than 249 items per page: `cf curl /v3/users?page=1\&per_page=250`. Would you like me to open a new issue or could we re-open this one?

**Comment by Gerg** [comment]
I'll re-open this issue to investigate.

**Comment by weymanf** [comment]
It's working with the latest version of capi "1.101.0". 

@robachmann could you please confirm?

**Comment by robachmann** [comment]
@weymanf thank you so much for taking care of this issue. It usually takes some weeks until upstream changes are available internally but I'm optimistic and think you can close this ticket in the meantime if you could reproduce the error in the first place and fixed it with this release. I would take the liberty of reopening the ticket if, contrary to expectations, problems should arise.

**Comment by weymanf** [comment]
Alright closing the ticket now. But please do re-open this ticket if you still have issues. It might be something strange happening with the actual users on UAA side.

**Comment by weymanf** [comment]
@robachmann Could you reproduce it on your current version and time your curls?

**Comment by weymanf** [comment]
We have finally reproduced it. We bumped up our users pass 2000+ and we curled with per_page 500.

@robachmann Looks like it's still an issue and we will be investigating this.

**Comment by Gerg** [comment]
This issue was resolved by 2e8fab8 and released in [capi-release 1.103.0](https://github.com/cloudfoundry/capi-release/releases/tag/1.103.0). Thanks for raising this issue!
